### PR TITLE
Create Block: Adds `--target-dir` flag to allow the tool to target where to scaffold

### LINKED ISSUE #44717: [Create Block] Allow explicit target for block creation
## What problem does this address?
I love that the `@wordpress/create-block` package allows the user to scaffold out a plugin for their blocks _and_ even has templates to allow developers to extend and create their own block template packages. ❤️ ❤️ 🤯 

I believe it is a high-priority need for developers to be able to create this type of structure in their plugin to support multiple blocks:

```
my-block-plugin/
    my-block-plugin.php
    src/
        block-one/
            block.json
        block-two/
            block.json
    build/
        block-one/
            block.json
        block-two/
            block.json
```

One _can_ technically accomplish this with `@wordpress/create-block` as it stands, but there are a few awkward steps to get there, which hinder its overall potential. 

To achieve that structure, you have to: `npx @wordpress/create-block my-block-plugin`, which outputs:

```
my-block-plugin/
    my-block-plugin.php
    src/
        block.json <-- no nesting in sub-directory
    build/
        block.json <-- no nesting in sub-directory
```

Now, you must manually move the `src/block.json` into a unique block directory: `src/block-one/block.json`, or delete it and run `@wordpress/create-block` again in another directory _and_ with the `--no-plugin` flag.

### Proposed solution: add `--target` flag

We add an optional flag for `--target`, which would allow developers to pass a block target within the `.src/`, e.g. `--target block-one/`, which would result in: `src/block-one/block.json`. If no `--target` then the `slug` should be the fallback used to create the target directory. Then, `npx @wordpress/create-block block-one` would result in the creation of `src/block-one/block.json` still.

These changes would allow builders to run the following and have a plugin: 

```
npx @wordpress/create-block my-block-plugin
cd my-block-plugin
rm -i src/*
npx @wordpress/create-block block-one --target block-one/ --no-plugin
```

and get the following:

```
my-block-plugin/
    my-block-plugin.php
    src/
        block-one/
            block.json
```

### PR DESCRIPTION
**Pull request by ryanwelcher**
## What?
This PR adds the `--target-dir` flag to allow users to target where the tool should scaffold either the entire plugin or the block files. This flag is not used when the tool is in `--no-plugin` mode

Closes #44717

## Why?
Being able to target where to scaffold the files is something that has been missing from the tool. By adding this flag extenders would be able to incorporate create-block into their scaffolds/workflow much easer.

For example, many projects use `wp-content` as the root directory and keep a package.json there. Currently, anyone wanting to use the create-block tool, would need to move the `plugins` directory and run it there. With this change, a script could be created to run the tool in place.

### Script
```
"scaffold-block": "npx @wordpress/create-block --target-dir plugins"
```

### Invoke
```
 npm run scaffold-block -- block-name
```
Another change here is now  when running the tool in `--no-plugin` mode, the resulting files are scaffolded into a sub-directory of the `folderName` property ( `./src` by default ) so now when adding a new block to an existing plugin the user no longer needs to move the the `folderName` directory. 

## How?
I am introducing a `targetDir `variable with the default value of `./` and prepending all of the paths with its value. 

## Testing Instructions
1. Checkout this branch
2. `cd` into `packages/create-block'
3. Run `node index.js test-no-target`
4. Confirm that a plugin is created at the root with the block files in a `src` directory
5. Run `node index.js test-with-target --target-dir target-directory`
6. Confirm that a plugin is created the `target-directory` folder with the block files in a `src` directory

### CODE REVIEW

File: packages/create-block/lib/index.js [hunk 1]
```diff
@@ -59,6 +59,7 @@ program
 	.option( '--wp-env', 'enable integration with `@wordpress/env` package' )
 	.option( '--no-plugin', 'scaffold only block files' )
 	.option( '--variant <variant>', 'the variant of the template to use' )
+	.option( '--target-dir <directory>', 'the variant of the template to use' )
```
Comment (bacoords) [inline packages/create-block/lib/index.js]: Just noting that this description still needs to be updated.

**Comment by bacoords** [comment]
I just tested this and noticed that if you combine it with the `--no-plugin` argument, it seems to ignore the `--target-dir` argument. 

Running `node index.js test-with-target --target-dir target-directory --no-plugin` I would've expected it to scaffold like one of these 
-  `target-directory/test-with-target` // where you have to put the `src` yourself since you're overwriting the dir
- `target-directory/src/test-with-target`
- `src/target-directory/test-with-target` 
- 
but instead it just puts it in the root. I'm not sure the best choice is but I'd lean towards the first one.

File: packages/create-block/lib/init-block.js [hunk 2]
```diff
@@ -38,8 +39,9 @@ async function initBlockJSON( {
 	info( 'Creating a "block.json" file.' );
 
 	const outputFile = plugin
-		? join( process.cwd(), slug, folderName, 'block.json' )
-		: join( process.cwd(), slug, 'block.json' );
+		? join( process.cwd(), targetDir, slug, folderName, 'block.json' )
+		: join( process.cwd(), folderName, slug, 'block.json' );
```
Comment (gziolo) [inline packages/create-block/lib/init-block.js]: From https://github.com/WordPress/gutenberg/pull/53781#issuecomment-1841144219:

> I just tested this and noticed that if you combine it with the --no-plugin argument, it seems to ignore the --target-dir argument.

It looks like `targetDir` should be used here as well.

**Comment by gziolo** [comment]
Thank you for testing @bacoords!

> Running `node index.js test-with-target --target-dir target-directory --no-plugin` I would've expected it to scaffold like one of these
> 
> * `target-directory/test-with-target` // where you have to put the `src` yourself since you're overwriting the dir
> * `target-directory/src/test-with-target`
> * `src/target-directory/test-with-target`
> 
> but instead it just puts it in the root. I'm not sure the best choice is but I'd lean towards the first one.

We could replicate [`wp scaffold plugin`](https://developer.wordpress.org/cli/commands/scaffold/plugin/) from WP-CLI. They have the `--dir` option that works as follows:

> [--dir=<dirname>]
> Put the new plugin in some arbitrary directory path. Plugin directory will be path plus supplied slug.

When using `--no-plugin` option, it would be up to the person using it to provide the correct path to the folder with sources of other blocks. Which is the first example included above.

Aside: I'm wondering how much we could read from the plugin when scaffolding another block. It could be useful to read the `namespace` and `textdomain`.

**Comment by bacoords** [comment]
> Aside: I'm wondering how much we could read from the plugin when scaffolding another block. It could be useful to read the namespace and textdomain.

True, although sometimes blocks are created in themes as well. I was hoping that once this was merged it would be easy to include a script like this in any project's package.json file:

```json
"create-block": "npx @wordpress/create-block --namespace=my-namespace --no-plugin --target-dir=src/blocks"
```

so that in the future you could run something like this from your project's root directory:

```
npm run create-block my-new-block
```

or

```
npm run create-block my-new-block -- --variant=dynamic
```

and it would generate it for you. So you could put the namespace and target-dir in your package.json.

**Comment by colorful-tones** [comment]
@ryanwelcher @gziolo and @bacoords thanks for pushing this forward!

I feel like I'm missing something here. This PR aims to address this issue: **Allow explicit target for block creation**. However, after utilizing the Testing Instructions I've merely created two plugins with blocks, and both versions of the plugin have the block stored in the usual `./src` directory. So, how are we allowing an explicit target for block creation? 🤔

**Comment by bacoords** [comment]
@colorful-tones Correct - currently this PR just lets you initiate the entire plugin in a different directory. Imagine you're in the root of your WP install, you could pass `--target-dir=wp-content/plugins` instead of having to `cd` to the plugins directory. 

But I think it becomes more useful if we could combine it with `--no-plugin` and then you can initiate an individual block somewhere other than the `/src` directory - such as in the [example above](https://github.com/WordPress/gutenberg/pull/53781#issuecomment-1843014639).

Happy to contribute/test here as needed.

**Comment by ryanwelcher** [comment]
This is all great feedback. I need to look at this PR again to resolve the conflicts but now I think the target should absolutely work in either mode.

**Comment by brettsmason** [comment]
I've tested this and can confirm it works as expected. I did wonder if this might help with https://github.com/WordPress/gutenberg/issues/56059 too, which it seems to. Though that still feels like an underlying bug. This would be a welcome addition though.

**Comment by gziolo** [comment]
@ryanwelcher, do you plan to refresh this PR?

**Comment by ryanwelcher** [comment]
> @ryanwelcher, do you plan to refresh this PR?

Yes! It's on my list to get back to.

**Comment by ryanwelcher** [comment]
I made some changes to this PR to make managing the paths much easier. Each function was generating the paths it needed internally which is really hard to keep up to date. When this command is run now, it will generate the path to the root and the block files based the `--no-plugin` and `--target-dir` flags.

There are now 4 cases that should work as follows:

1. `create-block new-plugin` -- scaffolds a new plugin called `new-plugin` in the location where the command was run.
2. `create-block new-block --no-plugin` -- scaffold only the block files in the location where the command was run.
3. `create-block another-plugin --target-dir plugins` -- scaffolds a new plugin called `another-plugin` inside a `plugins` folder.
4. `create-block another-block --no-plugin --target-dir new-plugin` -- scaffolds a new block called `another-block` inside the `new-plugin/src` directory.

This is going to need to be tested for all cases and hopefully with some custom external templates.

File: packages/create-block/lib/init-block.js (hunk 2 above)
Comment (ryanwelcher) [reply packages/create-block/lib/init-block.js]: I've changed how the paths are done so this is no longer relevant.

File: packages/create-block/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+-   Add `--target-dir` flag to allow indicating where resulting plugin will be scaffolded ([#53781](https://github.com/WordPress/gutenberg/pull/53781))
```
Comment (gziolo) [inline packages/create-block/CHANGELOG.md]:
```suggestion
### Enhancement

-   Add `--target-dir` flag to allow indicating where resulting plugin will be scaffolded ([#53781](https://github.com/WordPress/gutenberg/pull/53781))
```

**Comment by gziolo** [comment]
Excellent work, @ryanwelcher. I took it for a spin and tested it extensively. I really like the refactoring you did that introduced the `rootDirectory` property, which can now be used consistently in all places instead of constructing the path on the fly. In particular, it's essential as the path now depends on the new CLI option.

> There are now 4 cases that should work as follows:
> 
> 1. `create-block new-plugin` – scaffolds a new plugin called new-plugin in the location where the command was run.
> 1. `create-block new-block --no-plugin` – scaffold only the block files in the location where the command was run.
> 1. `create-block another-plugin --target-dir plugins` – scaffolds a new plugin called another-plugin inside a plugins folder.
> 1. `create-block another-block --no-plugin --target-dir new-plugin` – scaffolds a new block called another-block inside the new-plugin/src directory.

I love how flexible it is now to decide where the plugin is scaffolded. If the `--target-dir` is not provided then it defaults to the plugin's slug as before. Otherwise, you can put it anywhere you want:

```bash
> # slug "magic" as the directory name
> create-block magic
> # current directory
> create-block magic --target-dir .
> # "plugins" directory
>  create-block magic --target-dir plugins
> # "plugins/foo/boo" directory
>  create-block magic --target-dir plugins/foo/boo
```

Screenshot 2024-10-28 at 12 16 41

I'm not fully happy with how inconsistent your proposal is for the `--no-plugin` case. I would prefer it worked exactly the same for the scaffolded block. I proposed all the required changes in [32f0c16](https://github.com/WordPress/gutenberg/pull/53781/commits/32f0c16836eabb8bd37419d435d3faa69ac82395) to take into consideration.

Screenshot 2024-10-28 at 12 22 25

It would be up to the developer whether to scaffold the block folder inside the subfolder based on the slug name, or whether they would explicitly set the target location if they want it to be inside the proper location. 

I also improved the developer documentation and applied a few edits to colocate similar CLI options.

**Review by gziolo — APPROVED** [review]

**Comment by ryanwelcher** [comment]
> I'm not fully happy with how inconsistent your proposal is for the --no-plugin case. I would prefer it worked exactly the same for the scaffolded block.

I think this makes a lot of sense. At the time I was doing the refactor, I wanted the `--no-plugin` mode to be smarter than it needed to be. Thanks for the review and refactor!

**Comment by t-hamano** [comment]
In my environment, the slug in the PHP file doesn't seem to be replaced:

create-block

Perhaps because [the process for replacing the slug](https://github.com/WordPress/gutenberg/pull/53781/files#diff-39db4b394604ae4e798dbd0962bb0f97e337f174e36f3a29f6b5c616cc441745R88) has been moved?

**Comment by gziolo** [comment]
> In my environment, the slug in the PHP file doesn't seem to be replaced:

I think I broke it. Thank you for catching it. I’ll work on the fix tomorrow unless someone else is faster 👍

**Comment by ryanwelcher** [comment]
Fix is here - https://github.com/WordPress/gutenberg/pull/66528

**Pull request merged**